### PR TITLE
BFD-1878: Record timestamps for HAPI pointcuts in access.json

### PR DESCRIPTION
**Pull request by JeremyIR**
**JIRA Ticket:**
[BFD-1878](https://jira.cms.gov/browse/BFD-1878)

**User Story or Bug Summary:**

Record timestamps for HAPI pointcuts in access.json
---

### What Does This PR Do?

Measure time spent in BFD with HAPI interceptors and pointcuts.

### What Should Reviewers Watch For?

Is the time accurately measured/calculated for each point cut?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [x] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/TimerInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  @Hook(Pointcut.SERVER_INCOMING_REQUEST_PRE_HANDLED)
+  public void requestPreHandled(ServletRequestDetails theRequestDetails) {
+    // Record the time before the BFD Handler is called.
+    Long requestStartMilliseconds =
+        (Long) theRequestDetails.getServletRequest().getAttribute(BfdMDC.REQUEST_START_KEY);
+    if (requestStartMilliseconds != null) {
+      // convert requestStartMilliseconds to timestamp
+      Timestamp requestStart = new Timestamp(requestStartMilliseconds);
+      MDC.put(
+          "http_access.request.server_incoming_request_pre_handle_time", requestStart.toString());
```
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/TimerInterceptor.java:23]: we should move to the newer BfdMdc which came out after this POC

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/TimerInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import ca.uhn.fhir.interceptor.api.Pointcut;
+import ca.uhn.fhir.rest.server.servlet.ServletRequestDetails;
+import gov.cms.bfd.server.sharedutils.BfdMDC;
+import java.sql.Timestamp;
+import org.slf4j.MDC;
+
+@Interceptor
+public class TimerInterceptor {
+
+  @Hook(Pointcut.SERVER_INCOMING_REQUEST_PRE_HANDLED)
```
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/TimerInterceptor.java:14]: Please link to the hapi docs on these various interceptors and pointcuts

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/TimerInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import ca.uhn.fhir.interceptor.api.Hook;
+import ca.uhn.fhir.interceptor.api.Interceptor;
+import ca.uhn.fhir.interceptor.api.Pointcut;
+import ca.uhn.fhir.rest.server.servlet.ServletRequestDetails;
+import gov.cms.bfd.server.sharedutils.BfdMDC;
+import java.sql.Timestamp;
+import org.slf4j.MDC;
+
+@Interceptor
+public class TimerInterceptor {
```
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/TimerInterceptor.java:12]: missing the javadocs in here

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/TimerInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  }
+
+  @Hook(Pointcut.SERVER_OUTGOING_RESPONSE)
+  public void serverOutgoingResponse(ServletRequestDetails theRequestDetails) {
+    // Record the response duration.
+    Long requestStartMilliseconds =
+        (Long) theRequestDetails.getServletRequest().getAttribute(BfdMDC.REQUEST_START_KEY);
+    if (requestStartMilliseconds != null)
+      MDC.put(
+          "http_access.response.server_outgoing_response_duration",
```
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/TimerInterceptor.java:34]: we're going to need to add the two new columns to the insights terraform

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/TimerInterceptor.java (hunk 1 above)
Comment (JeremyIR) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/TimerInterceptor.java:23]: Updating

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/V2Server.java [hunk 5]
```diff
@@ -113,5 +113,7 @@ protected void initialize() throws ServletException {
     // Default to XML and pretty printing.
     setDefaultResponseEncoding(EncodingEnum.JSON);
     setDefaultPrettyPrint(false);
+
+    registerInterceptor(new TimerInterceptor());
```
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/V2Server.java:117]: a comment here that this registers hapi interceptors to capture metrics on the timing characteristics of the response handling would be great (same comment on V1)

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/TimerInterceptor.java (hunk 4 above)
Comment (JeremyIR) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/TimerInterceptor.java:34]: Adding

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/V2Server.java (hunk 5 above)
Comment (JeremyIR) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/V2Server.java:117]: Updating

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/TimerInterceptor.java (hunk 1 above)
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/TimerInterceptor.java:23]: I would like to change the name of this to make it more clear where it came from. proposing 'hapi. server_incoming_request_pre_handle_time' (similar comment on the other one)

Comment (JeremyIR) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/TimerInterceptor.java:23]: Sounds good

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/TimerInterceptor.java (hunk 2 above)
Comment (JeremyIR) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/TimerInterceptor.java:24]: Updated

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/TimerInterceptor.java (hunk 3 above)
Comment (JeremyIR) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/TimerInterceptor.java:17]: Updated

**Comment by christrotteradhoc** [comment]
Minor comment -- we should change the description of the story and the commit/pr title to reflect the change that we made to the AC. "Record timestamps for HAPI pointcuts in access.json"

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/TimerInterceptor.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+ * https://hapifhir.io/hapi-fhir/docs/interceptors/server_pointcuts.html
+ */
+@Interceptor
+public class TimerInterceptor {
+
+  /** Pointcut to log time when a request is pre-processed. */
+  @Hook(Pointcut.SERVER_INCOMING_REQUEST_PRE_PROCESSED)
+  public void requestPreProcessed() {
+    // log current instance to MDC
+    Timestamp currentTime = new Timestamp(System.currentTimeMillis());
```
[the remark this answers is not in the document]
Comment (JeremyIR) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/TimerInterceptor.java:22]: Lets go with milliseconds then.

**Review by christrotteradhoc — APPROVED** [review]

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/TimerInterceptor.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+@Interceptor
+public class TimerInterceptor {
+
+  /** Pointcut to log time in milliseconds when a request is pre-processed. */
+  @Hook(Pointcut.SERVER_INCOMING_REQUEST_PRE_PROCESSED)
+  public void requestPreProcessed() {
+    // log current instance to MDC
+    Long currentTime = System.currentTimeMillis();
+    BfdMDC.put(
+        "hapi.server_incoming_request_pre_process_time_in_millis", Long.toString(currentTime));
```
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/TimerInterceptor.java:23]: sorry didn't see this before -- I think these should be timestamp_in_millis (time could imply duration to some). Thanks for making the other changes!

Comment (JeremyIR) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/TimerInterceptor.java]: Updated

**Review by christrotteradhoc — APPROVED** [review]

**Review by cbrunefearless — APPROVED** [review]
lgtm

**Pull request merged**